### PR TITLE
Move allowed_rolldev here from sparkles and update pitch/roll table

### PR DESCRIPTION
**Pull request by jeanconn**
## Description

Move allowed_rolldev here from sparkles

Also use updated pitch/roll table. The values come from the update file with a restriction to 17deg starting at pitch 125 and shared with MPCWG in March 2022 (MPCWG/MeetingNotes2022x03x24). That file has signed/positive and negative roll ranges but all the magnitudes are the same for each sign for each row, so reduced the table to the unsigned version.

## Testing

- [x] Passes unit tests on MacOS
- [x] Functional testing - new simple test added that is hardcoded for the current pitch/roll table.

### Pitch / rolldev data

(this includes a point added at pitch 180 defined as 0 deg allowed off-nominal-roll)

pitch_roll

### Install test

This was successful.
```
python setup.py install --user
$ cd ~
$ ipython
import Ska.Sun
print(Ska.Sun.__file__)  # version in ~/.local
Ska.Sun.test('-v')  # pass
```

### CODE REVIEW

File: Ska/Sun/sun.py [hunk 1]
```diff
@@ -2,14 +2,29 @@
...
+from astropy.table import Table
 
 from Quaternion import Quat
 from chandra_aca.transform import radec_to_eci
 from Chandra.Time import DateTime
-from math import cos, sin, acos, atan2, asin, pi, radians, degrees
-import numpy as np
 import Ska.quatutil
 
+ROLL_TABLE = Table.read(str(Path(__file__).parent / 'data' / 'pitch_rolldev.csv'),
```
Comment (taldcroft) [inline Ska/Sun/sun.py:15]: Do this as a `ska_helpers.util.LazyVal` instance so the table is read only on demand. 

I'm being more conscious of avoiding side-effects of importing modules, and `LazyVal` is an easy way to do that. It does have a syntactic disadvantage of having to write `ROLL_TABLE.val` later, but in this context that is not a problem.

File: Ska/Sun/sun.py [hunk 2]
```diff
@@ -2,14 +2,29 @@
...
+                        format='ascii.basic',
+                        guess=False, delimiter=',')
+
+
+def allowed_rolldev(pitch):
+    """Get allowed roll deviation (off-nominal roll) for the given ``pitch``.
+    :param pitch: Sun pitch angle (deg)
+    :returns: Roll deviation (deg)
+    """
+    idx = np.searchsorted(ROLL_TABLE['pitch'], pitch, side='right')
```
Comment (taldcroft) [inline Ska/Sun/sun.py:25]: I am not sure this is correct for the new table, which is no longer done as a stairstep but rather samples values. I'd say that `np.interp()` is more appropriate. This should (by default) give the right behavior for pitch values outside the table range.

File: Ska/Sun/sun.py [hunk 3]
```diff
@@ -2,14 +2,29 @@
...
+    return ROLL_TABLE['rolldev'][idx - 1]
```
Comment (taldcroft) [inline Ska/Sun/sun.py:26]: There is a problem here with the `idx - 1`:
```
In [1]: import Ska.Sun
Sk
In [2]: Ska.Sun.__file__
Out[2]: '/Users/aldcroft/git/Ska.Sun/Ska/Sun/__init__.py'

In [3]: Ska.Sun.allowed_rolldev(40)
Out[3]: 18.7487718079019  <== corresponds to the [-1] entry
```

File: Ska/Sun/tests/test_sun.py [hunk 4]
```diff
@@ -3,7 +3,15 @@
...
-                   off_nominal_roll, position, pitch as sun_pitch)
+                   off_nominal_roll, position, allowed_rolldev,
+                   pitch as sun_pitch)
+
+
+def test_allowed_rolldev():
+    dev1 = allowed_rolldev(135)
+    assert np.isclose(dev1, 13.979)
+    dev2 = allowed_rolldev(138)
+    assert np.isclose(dev2, 14.516)
```
Comment (taldcroft) [inline Ska/Sun/tests/test_sun.py:15]: Add a test for 0, 40, 180.0 and 181.0 deg.

File: setup.py [hunk 5]
```diff
@@ -14,7 +14,7 @@
       setup_requires=['setuptools_scm', 'setuptools_scm_git_archive'],
       zip_safe=False,
       packages=['Ska', 'Ska.Sun', 'Ska.Sun.tests'],
-      package_data={},
+      package_data={'Ska.Sun': ['data/*csv']},
```
Comment (taldcroft) [inline setup.py:17]: It's probably worth making this file smaller, esp. since it will eventually have the + and - columns.
```
In [10]: t = Table.read('Ska/Sun/data/pitch_rolldev.csv')
In [11]: t['pitch'] = t['pitch'].astype(np.float32)
In [12]: t['rolldev'] = t['rolldev'].astype(np.float32)
In [13]: t.write('junk.fits.gz', overwrite=True)
```
This compresses the file from 430k to 43k.

File: Ska/Sun/sun.py (hunk 1 above)
Comment (taldcroft) [reply Ska/Sun/sun.py:15]: The `str` around `Path` is not needed.

**Review by taldcroft — CHANGES_REQUESTED** [review]

File: Ska/Sun/sun.py (hunk 2 above)
Comment (jeanconn) [reply Ska/Sun/sun.py:25]: "I am not sure this is correct for the new table, which is no longer done as a stairstep but rather samples values." The table we were using was the 278 line table (not the 62 line matlab constraints stairstep..)  but I still thought it was defined as a stairstep.  Maybe we should check?  Is the right behavior for pitch values outside the table range 0 or error?

File: Ska/Sun/sun.py (hunk 3 above)
Comment (jeanconn) [reply Ska/Sun/sun.py:26]: Good catch.  I did not test that the previous routine was working for the corner cases.

File: setup.py (hunk 5 above)
Comment (jeanconn) [reply setup.py:17]: I disagree on that one.

File: Ska/Sun/sun.py (hunk 1 above)
Comment (jeanconn) [reply Ska/Sun/sun.py:15]: Regarding the "str" around "Path".  What is the nicest way to make this work with format='ascii.basic'?  ```     20     """
     21     file = Path(__file__).parent / 'data' / 'pitch_rolldev.csv'
---> 22     ROLL_TABLE = Table.read(file,
     23                             format='ascii.basic',
     24                             guess=False, delimiter=',')

/proj/sot/ska3/flight/lib/python3.8/site-packages/astropy/table/connect.py in __call__(self, *args, **kwargs)
     59         descriptions = kwargs.pop('descriptions', None)
     60 
---> 61         out = registry.read(cls, *args, **kwargs)
     62 
     63         # For some readers (e.g., ascii.ecsv), the returned `out` class is not

/proj/sot/ska3/flight/lib/python3.8/site-packages/astropy/io/registry.py in read(cls, format, cache, *args, **kwargs)
    518 
    519         reader = get_reader(format, cls)
--> 520         data = reader(*args, **kwargs)
    521 
    522         if not isinstance(data, cls):

/proj/sot/ska3/flight/lib/python3.8/site-packages/astropy/io/ascii/connect.py in io_read(format, filename, **kwargs)
     16         format = re.sub(r'^ascii\.', '', format)
     17         kwargs['format'] = format
---> 18     return read(filename, **kwargs)
     19 
     20 

/proj/sot/ska3/flight/lib/python3.8/site-packages/astropy/io/ascii/ui.py in read(table, guess, **kwargs)
    355             fast_reader_rdr = get_reader(**fast_kwargs)
    356             try:
--> 357                 dat = fast_reader_rdr.read(table)
    358                 _read_trace.append({'kwargs': copy.deepcopy(fast_kwargs),
    359                                     'Reader': fast_reader_rdr.__class__,

/proj/sot/ska3/flight/lib/python3.8/site-packages/astropy/io/ascii/fastbasic.py in read(self, table)
    107         self.kwargs['fast_reader'] = fast_reader
    108 
--> 109         self.engine = cparser.CParser(table, self.strip_whitespace_lines,
    110                                       self.strip_whitespace_fields,
    111                                       delimiter=self.delimiter,

astropy/io/ascii/cparser.pyx in astropy.io.ascii.cparser.CParser.__cinit__()

astropy/io/ascii/cparser.pyx in astropy.io.ascii.cparser.CParser.setup_tokenizer()

TypeError: Input "table" must be a file-like object, a string (filename or data), or an iterable```

File: Ska/Sun/tests/test_sun.py (hunk 4 above)
Comment (jeanconn) [reply Ska/Sun/tests/test_sun.py:15]: And what values are the right answers for 180 and 181 (both greater than the last entry in the table).

File: Ska/Sun/sun.py (hunk 1 above)
Comment (taldcroft) [reply Ska/Sun/sun.py:15]: Argh, sorry about the `str`. I thought we had full support for `Path` everywhere, but apparently not. So just put the `str()` back. If you are feeling energetic you can open a bug report on astropy, this is definitely a bug.

File: Ska/Sun/sun.py (hunk 2 above)
Comment (taldcroft) [reply Ska/Sun/sun.py:25]: After that long slack conversation I think the approach below could work.  It is close but always conservative by choosing the minimum of points to the left and to the right:
```
idx0 = np.searchsorted(ROLL_TABLE['pitch'], pitch, side='left')
idx1 = np.searchsorted(ROLL_TABLE['pitch'], pitch, side='right')
# These need some bound checking at the edges
val0 = ROLL_TABLE['rolldev'][idx0]
val1 = ROLL_TABLE['rolldev'][idx1]
out = np.minimum(val0, val1)  # works even for a vector input for `pitch`
return out
```

File: Ska/Sun/tests/test_sun.py (hunk 4 above)
Comment (taldcroft) [reply Ska/Sun/tests/test_sun.py:15]: The right answers are 0.0.

File: Ska/Sun/sun.py (hunk 2 above)
Comment (taldcroft) [reply Ska/Sun/sun.py:25]: >  Is the right behavior for pitch values outside the table range 0 or error?

The right answer is 0.0.

File: setup.py (hunk 5 above)
Comment (taldcroft) [reply setup.py:17]: This isn't really a file that is meant for human consumption. What is the advantage of CSV?  FITS is also faster to load.

Comment (jeanconn) [reply setup.py:17]: I just disagree with the "probably worth".  Granted a trivial cost to making the file smaller (and adding documentation someplace about how to make the file), but the price is losing transparency that includes github diffs (as in the FOT can see what we are using from github).

Comment (taldcroft) [reply setup.py:17]: OK. I don't think that the FOT or anyone will care about text diffs of a huge data file. That's not a useful way to see what has changed. If there is a diff then we would want a plot of some sort. If being complete with documentation, then you would start from referencing an email and document the step of translating that excel file to CSV. From there it is 3 lines of code to write out as FITS.

I just don't want to needlessly bloat this package to (eventually) nearly a Mb.

Comment (jeanconn) [reply setup.py:17]: OK.  I figured we'd just add a comment to the top of the CSV referencing the email and the original file name and it would be self-documenting, but fine if you want to save 400k compress it and stick the notes someplace else.  Where should those notes belong... just in the package readme?

Comment (taldcroft) [reply setup.py:17]: You can make a Python script `make_rolldev_file.py` at the top level in the repo. There you can put the actual code and documentation about provenance.

(In bigger repos we might hide that file in a `dev` or `utils` directory, but I think top-level is fine here).

Comment (jeanconn) [reply setup.py:17]: OK.  Though note that I'm not planning to set code to remake the csv we have now (which I just converted from xlsx or whatever using google sheets and then reduced to one column with astropy table).

Comment (taldcroft) [reply setup.py:17]: Sounds good.

File: Ska/Sun/tests/test_sun.py (hunk 4 above)
Comment (jeanconn) [reply Ska/Sun/tests/test_sun.py:15]: I thought the right answer for 180 was still technically the same as the last row in the file (close to but not equal to 180), but suppose that's in the noise.

Comment (taldcroft) [reply Ska/Sun/tests/test_sun.py:15]: Whether it is 0. for pitch >= 180.0 or pitch > 180 doesn't matter, but certainly for pitch > 180 it should be 0. Whatever ends up being the cleanest implementation can drive the answer there. I would think about adding a value of 0 to the table at 180 so the code is cleaner, but whatever works out.

File: Ska/Sun/sun.py (hunk 2 above)
Comment (jeanconn) [reply Ska/Sun/sun.py:25]: I don't quite understand the bounds checking here, but how do you want to handle the >= 180 cases then?
I'd probably just stick some logic to define outside of bounds as 0 but...


```
In [6]: allowed_rolldev(180)
---------------------------------------------------------------------------
IndexError                                Traceback (most recent call last)
<ipython-input-6-c0f6f350c041> in <module>
----> 1 allowed_rolldev(180)

~/git/Ska.Sun/Ska/Sun/sun.py in allowed_rolldev(pitch)
     30     idx1 = np.searchsorted(ROLL_TABLE.val['pitch'], pitch, side='right')
     31     # These need some bound checking at the edges
---> 32     val0 = ROLL_TABLE.val['rolldev'][idx0]
     33     val1 = ROLL_TABLE.val['rolldev'][idx1]
     34     out = np.minimum(val0, val1)  # works even for a vector input for `pitch`

astropy/table/_column_mixins.pyx in astropy.table._column_mixins._ColumnGetitemShim.__getitem__()

astropy/table/_column_mixins.pyx in astropy.table._column_mixins.base_getitem()

astropy/table/_column_mixins.pyx in astropy.table._column_mixins.column_getitem()

IndexError: index 13340 is out of bounds for axis 0 with size 13340
```

Comment (jeanconn) [reply Ska/Sun/sun.py]: And did you want to support vector pitch or not?

Comment (taldcroft) [reply Ska/Sun/sun.py]: Yes on vector pitch.

Comment (taldcroft) [reply Ska/Sun/sun.py]: This might work:
```
idx_max =  len(ROLL_TABLE.val) - 1
idx0 = np.clip(idx0, 0, idx_max)
idx1 = np.clip(idx1, 0, idx_max)
```

Comment (jeanconn) [reply Ska/Sun/sun.py]: If you meant the function as a whole would look like 
```
    idx0 = np.searchsorted(ROLL_TABLE.val['pitch'], pitch, side='left')
    idx1 = np.searchsorted(ROLL_TABLE.val['pitch'], pitch, side='right')
    idx_max =  len(ROLL_TABLE.val) - 1
    idx0 = np.clip(idx0, 0, idx_max)
    idx1 = np.clip(idx1, 0, idx_max)
    # These need some bound checking at the edges
    val0 = ROLL_TABLE.val['rolldev'][idx0]
    val1 = ROLL_TABLE.val['rolldev'][idx1]
    return np.minimum(val0, val1) 
```
It looks like for pitch >= 180 idx0 and idx1 both end up set at the last element so they stick at 18.748.  I was thinking about something else, but if you want to play with this that's fine.

File: Ska/Sun/tests/test_sun.py (hunk 4 above)
Comment (jeanconn) [reply Ska/Sun/tests/test_sun.py:18]: I cheated and just added the value at 180 in the code for now.  I'm not sure if that may be cleaner overall than adding a value to the data file.

File: Ska/Sun/sun.py (hunk 2 above)
Comment (jeanconn) [reply Ska/Sun/sun.py]: If you meant this in combination with adding another row to the data file or the table with pitch(180) = 0, that works.

File: Ska/Sun/tests/test_sun.py [hunk 6]
```diff
@@ -3,7 +3,18 @@
 import numpy as np
 from Quaternion import Quat
 from ..sun import (apply_sun_pitch_yaw, get_sun_pitch_yaw, nominal_roll,
-                   off_nominal_roll, position, pitch as sun_pitch)
+                   off_nominal_roll, position, allowed_rolldev,
+                   pitch as sun_pitch)
+
+
+def test_allowed_rolldev():
+    assert np.isclose(allowed_rolldev(135), 13.979)
```
Comment (taldcroft) [inline Ska/Sun/tests/test_sun.py:11]: There should be one case which is fine-tuned to land between a steep transition (at the throat) to prove that the `minimum` business is working.

Comment (taldcroft) [reply Ska/Sun/tests/test_sun.py:11]: Also add a test with an array input. You can just repeat all of the scalar values.

File: setup.py [hunk 7]
```diff
@@ -14,7 +14,7 @@
       setup_requires=['setuptools_scm', 'setuptools_scm_git_archive'],
       zip_safe=False,
       packages=['Ska', 'Ska.Sun', 'Ska.Sun.tests'],
-      package_data={},
+      package_data={'Ska.Sun': ['data/*fits.gz']},
```
Comment (taldcroft) [inline setup.py:17]: I never trust myself with packaging updates like this. Best to do a quick functional install test. I usually do:
```
python setup.py install --user
$ cd ~
$ ipython
import Ska.Sun
Ska.Sun.test()
exit()
$ rm -rf ~/.local  # assuming you keep this clean like I do.
```

File: Ska/Sun/sun.py (hunk 2 above)
Comment (jeanconn) [reply Ska/Sun/sun.py]: Right, it looks like it mostly worked, but side='left' really doesn't give you the index left of the the value you want... it still gives you the index that the value would have.  Subtracting 1 from the side='right' answer to get idx0 might be a little closer?  I've updated for that for now.

Comment (taldcroft) [reply Ska/Sun/sun.py]: Good catch with the testing. I had fallen into the trap of thinking about the "left" and "right" values incorrectly, even though I should know better. Bah. The solution of subtracting 1 is correct!

**Comment by taldcroft** [comment]
I thought the team had decided to set the start of the FSS restriction at a pitch of 125? This is what is in the file here in Ska.Sun:
image

**Comment by jeanconn** [comment]
Good point.   I probably should have done this as 3 PRs so I'd have kept better track of file versions.

File: Ska/Sun/tests/test_sun.py [hunk 8]
```diff
@@ -3,7 +3,31 @@
...
+               [138, 14.516],
+               [0, 0],
+               [40, 0],
+               [179.9, 18.748772],
+               [179.997, 0],
+               [180, 0],
+               [181, 0],
+               [85.49229, 13.677669],
+               [85.52, 18.756727],
+               [125.99, 18.748772],
```
Comment (taldcroft) [inline Ska/Sun/tests/test_sun.py:23]: These can be 124.99 and 125.0 pitch now.

**Comment by taldcroft** [comment]
I don't follow the 3 PRs remark. This PR seems good, though you'll want to squash down to one commit at the end.

File: make_rolldev_file.py [hunk 9]
```diff
@@ -0,0 +1,13 @@
+# For the pitch/roll table.  I used google sheets to save the xlsx file as
+# a CSV (after adding headers: pitch, negrolldev, rolldev).
```
Comment (taldcroft) [inline make_rolldev_file.py:2]: Better to read the excel file directly.
```
In [12]: df = pandas.read_excel('/Users/aldcroft/Downloads/pitch_roll_2022.xlsx'
    ...: , names=['pitch', 'rolldev_minus', 'rolldev_plus'])

In [13]: df
Out[13]: 
            pitch  rolldev_minus  rolldev_plus
0       46.612288      -0.060745      0.060745
1       46.622288      -0.116114      0.116114
2       46.632288      -0.171483      0.171483
3       46.642288      -0.224661      0.224661
4       46.652288      -0.274136      0.274136
...           ...            ...           ...
13334  179.952288     -18.748772     18.748772
13335  179.962288     -18.748772     18.748772
13336  179.972288     -18.748772     18.748772
13337  179.982288     -18.748772     18.748772
13338  179.992288     -18.748772     18.748772

[13339 rows x 3 columns]

In [14]: t = Table.from_pandas(df)

In [15]: t
Out[15]: 
<Table length=13339>
      pitch           rolldev_minus         rolldev_plus   
     float64             float64              float64      
------------------ -------------------- -------------------
 46.61228812838714 -0.06074530362457799 0.06074530362457799
 46.62228812838715 -0.11611406766359549 0.11611406766359549
46.632288128387145  -0.1714828317025736  0.1714828317025736
 46.64228812838714 -0.22466119433237589 0.22466119433237589
 46.65228812838714 -0.27413563619862313 0.27413563619862313
 46.66228812838715  -0.3236100780649055  0.3236100780649055
               ...                  ...                 ...
179.93228812838714  -18.748771807903847  18.748771807903847
179.94228812838713  -18.748771807903516  18.748771807903516
179.95228812838715  -18.748771807903186  18.748771807903186
179.96228812838714  -18.748771807902855  18.748771807902855
179.97228812838713  -18.748771807902525  18.748771807902525
179.98228812838715  -18.748771807902195  18.748771807902195
179.99228812838714  -18.748771807901864  18.748771807901864
```

Comment (jeanconn) [reply make_rolldev_file.py:2]: It is better, but I don't see the point of the rework at this point.

Comment (taldcroft) [reply make_rolldev_file.py:2]: Well at least document the better way for future updates, which are 100% certain to happen.

**Comment by jeanconn** [comment]
For me, it would actually have been 4 PRs.  One to move the code here from sparkles.  One to update to the reduced roll range data file.  One to update to the modified-at-126 file.  One to update to the modified-at-125 file.  I got confused about which version we had and what we wanted because there was all of this combining-into-on-step process.

File: setup.py (hunk 7 above)
Comment (jeanconn) [reply setup.py:17]: And are you saying in the edited description that you've done the install test?

Comment (jeanconn) [reply setup.py:17]: OK, well I've done it too so it can stay.

File: make_rolldev_file.py (hunk 9 above)
Comment (jeanconn) [reply make_rolldev_file.py:2]: Good idea.  Thanks!

**Comment by taldcroft** [comment]
I think each of the 4 PR's you mentioned are more at the level of single commits. I had no trouble following this and your single PR as a reviewer. In the end the first 3 PR's you made would be abandoned since I don't want commits of the wrong/intermediate data files in the repo.

**Review by taldcroft — APPROVED** [review]
Looks good!

**Comment by jeanconn** [comment]
Thanks for all the help with this one!
